### ISSUE TITLE
Issues with KnowPoses, Frustum and High Residuals

### ISSUE THREAD

**Issue by belyeng93** [openMVG/openMVG#1858]
I am using `openMVG_main_ComputeStructureFromKnownPoses `  to compute a Sfm knowing the intrinsic and extrinsic parameters like @treyfortmuller in the #1013 issue .
I've a rotating arm mounted into an optical encoder and fixed into a another arm placed in a desk.  The camera, a global shutter arducam, is mounted perpendicular to the rotating arm as shown in the following picture that represent the entire structure:
![box](https://user-images.githubusercontent.com/9927772/108500464-18c05300-72b0-11eb-989d-0b50f2f11720.PNG)

The HW is handled by some ROS nodes that provide us, using tf2 lib, the position of the camera at each rotation as the following image:

![posizione](https://user-images.githubusercontent.com/9927772/108504387-14973400-72b6-11eb-93be-0e58b8e7ff45.PNG)

As you can see, the world has the same notation of the camera and the rotation is around the Y axis.

Tf2 lib gives us the position of the camera in the world coordinate system (![equation1](https://latex.codecogs.com/gif.latex?H%5E%7Bcamera%7D_%7Bworld%7D)) and, at the same time, the word position from the camera coordinate system (Extrinsic Parameters ) ![equation2](https://latex.codecogs.com/gif.latex?%5Cinline%20%5BR%7Ct%5D) = ![equation3](https://latex.codecogs.com/gif.latex?%5Cinline%20H%5E%7Bcamera%7D_%7Bworld%7D) or in other hands 

![equation4](https://latex.codecogs.com/gif.latex?%5Cinline%20P%5E%7Bcamera%7D%20%3D%20H%5E%7Bcamera%7D_%7Bworld%7D%20*%20P%5E%7Bworld%7D)

**Question 1**: If I have understood correctly, the poses in the sfm_data.JSON are:

- Rotation: the rotation R of the extrinsixc parameters;
- Position: the camera position in the world coordinate system or in other hands ![equation](https://bit.ly/3s7LaZN);

**Question 2**: If i set the extrinsic rotation R as an Identity matrix (no rotation) and C = [0,0,0] (no traslation) i have the following frustum, as in the following image, with the Y-down, X-right and Z-forward it's correct (right handed)? Because in some issues, like #1756 and #1785, is mentioned a left-handed system.

![id](https://user-images.githubusercontent.com/9927772/108518776-3e0d8b00-72c9-11eb-9016-bcc62b6d56e7.PNG)

**Question 3**:  This is the really question. After some datsets and multiple HW  and SW changes to obtain the best precision of the pose, I still can't get one good poincloud using  `openMVG_main_ComputeStructureFromKnownPoses ` or  `openMVG_main_IncrementalSfm2 ` with EXISTING POSE and POSE PRIORS as shown following.

![knowposes join](https://user-images.githubusercontent.com/9927772/108523077-cd1ca200-72cd-11eb-889e-c59801c43e97.PNG)

Obviously `openMVG_main_IncrementalSfm2 ` , `openMVG_main_IncrementalSfm ` and `openMVG_main_GlobalSfm` give us a good pointcloud as shown following.

Global:

![GLOBALJOIN](https://user-images.githubusercontent.com/9927772/108523179-e9b8da00-72cd-11eb-968c-9a2532d55538.PNG)

Incremental:

![incrementaljoin](https://user-images.githubusercontent.com/9927772/108523215-f2111500-72cd-11eb-8820-5b20b8ee6c7d.PNG)

To get more points in the know poses algorithm i've set an high residual threshold value, but the quality of the result is very bad and always i have this kind of residual graph:

r = 10

![residual](https://user-images.githubusercontent.com/9927772/108531146-ab73e880-72d6-11eb-832f-79d8b64ac6ad.PNG)

r = 4

![r4](https://user-images.githubusercontent.com/9927772/108532167-d9a5f800-72d7-11eb-8045-b279379c3596.PNG)

@pmoulon what do you think about this problem? Is this a pose problem or do I make mistakes somewhere else?

Other things I think might help are:

- Sometimes some parts of the same object are plotted in different planes as in the following image:

![più piani](https://user-images.githubusercontent.com/9927772/108530692-2688cf00-72d6-11eb-954b-bfb849a3c933.PNG)

- A few times if I remove on image i get a good result.

- I use meters for the poses

**Question 4**:  getting a high residual value due to a positioning error?
 

Thanks in advance

**Comment by pmoulon** [comment]
Hello,

Q1: You are right you have the right convention R and C

Q2: Y Down, Z Front is right for OpenMVG.

Q3: I would advise you to compute the transform between your manual poses and the one found by SfM (https://github.com/openMVG/openMVG/blob/develop_2.0/src/software/SfM/main_evalQuality.cpp) evalQuality can now load two sfm_data and compare them

I think we see two mode in the histogram due to the circular motion of the camera and the fact that your scene has some repetitive structure (and so it makes duplicate structure appears).

I would advise you to take two capture at a different distance if you can and it could help to remove those wrong matches.

BTW: Thank you for taking the time to carefully illustrate what you tried. Appreciated!

**Comment by pmoulon** [comment]
Closing due to inactivity. Feel free to reopen

**Comment by muskie82** [comment]
Would appreciate it if you share an update on this.
I also had a problem with using known poses. In my case, I provide GT poses on a synthetic dataset (Replica dataset) but IncrementalSfM2 fails, while IncrementalSfM without known poses works quite well.  

Other people also seem to meet this issue.
#1420  #1851 #1756

**Comment by pmoulon** [comment]
@muskie82 Can you share you data importer, something must be wrong in the import?

**Comment by manurare** [comment]
Hey @muskie82. I am having similar problems. How do you represent the poses? I guess if you got them from Blender (or a similar software) you need to make sure the poses are from openMVG camera to world openMVG. The latter I am not sure what is the coordinate system. 

During BA without known poses I think the world is initialized to Y up, X right since one random camera is taken as reference and set to identity pose. But I don't know what happens if we input our own poses.

**Comment by muskie82** [comment]
Thanks for your reply.
I finally made it work... in my case, the problem was the coordinate system.
Replica (or other SLAM-related datasets) typically represents the gt trajectory as a camera poses in the world coordinate system (say, `T_cam_world`). However, after I carefully read the OpenMVG code, I noticed that I have to invert the gt poses. GT camera poses are to be represented as the origin of the world coordinate system in the camera local coordinate system (`T_world_cam`). 
After inverting the camera pose matrix, it works.... hope this helps you as well.
